src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -554,36 +556,89 @@ private void CheckForHttp11ConnectionInjection()
+        private void ProcessHttp11RequestQueue(HttpConnection? connection)
         {
+            // Try to signal a single request that is waiting for a connection.
+            // We're expecting this method to be called enough times to handle all requests.
             while (true)
             {
+                HttpConnectionWaiter<HttpConnection>? waiter = null;
+
                 lock (SyncObj)
                 {
-                    _usedSinceLastCleanup = true;
+                    // This assert message may be misleading due to a race condition (we're reading Count twice).
+                    Debug.Assert(_associatedHttp11ConnectionCount >= _http11Connections.Count + _pendingHttp11ConnectionCount,
+                        $"Expected {_associatedHttp11ConnectionCount} >= {_http11Connections.Count} + {_pendingHttp11ConnectionCount}");
+                    Debug.Assert(_associatedHttp11ConnectionCount <= _maxHttp11Connections,
+                        $"Expected {_associatedHttp11ConnectionCount} <= {_maxHttp11Connections}");
+                    Debug.Assert(_associatedHttp11ConnectionCount >= _pendingHttp11ConnectionCount,
+                        $"Expected {_associatedHttp11ConnectionCount} >= {_pendingHttp11ConnectionCount}");
 
-                    int availableConnectionCount = _availableHttp11Connections.Count;
-                    if (availableConnectionCount > 0)
+                    if (_http11RequestQueue.Count != 0)
                     {
-                        // We have a connection that we can attempt to use.
-                        // Validate it below outside the lock, to avoid doing expensive operations while holding the lock.
-                        connection = _availableHttp11Connections[availableConnectionCount - 1];
-                        _availableHttp11Connections.RemoveAt(availableConnectionCount - 1);
+                        if (connection is not null || _http11Connections.TryDequeue(out connection))
+                        {
+                            bool success = _http11RequestQueue.TryDequeueWaiter(this, out waiter);
+                            Debug.Assert(success == waiter is not null);
+                        }
                     }
-                    else
+
+                    // Update the empty queue flag now.
+                    _http11RequestQueueIsEmptyAndNotDisposed = _http11RequestQueue.Count == 0 && !_disposed;
+
+                    if (waiter is null)
                     {
-                        // No available connections. Add to the request queue.
-                        waiter = _http11RequestQueue.EnqueueRequest(request);
+                        // We didn't find a waiter to signal, or there were no connections available.
+
+                        if (connection is not null)
+                        {
+                            // A connection was provided to this method, or we rented one from the pool.
```

Comment:
Related to my comment above, I'm not clear how we could end up in a situation where we rented one from the pool but waiter is null. Won't we only rent one from the pool if we're able to successfully get a waiter? I don't think that affects the logic here, just the comment.